### ISSUE TITLE
Removal of ComicInfo.xml during conversion

### ISSUE THREAD

**Issue by 5h4k3** [binarynonsense/comic-book-reader#26]
During a conversion, ACBR remove everything that is not a image. Even the ComicInfo.xml.

If the archive contains folders which contain the images, after the conversion the output is an archive with the converted images directly inside. This happens every time with one but even with two, three or four nested folders. Maybe more, I haven't tested with more than four. That's cool.

If the archive contains any files that is not a image, the same thing happens. All these files are removed from the final result. That's cool too, except for one thing. I think any comic lover would agree that it will be better to keep the ComicInfo.xml but delete everything else.

In conclusion, this may be a deliberate design but it has the side effect of making me use the conversion feature only for my new comics. I proceed differently to convert the ones that are already in my library because I don't want to lose all the work already done to collect the metadata.

Anyway i wanted to thank you for this awesome app.

**Comment by binarynonsense** [comment]
Thanks for your input! This started as a comic book reader (later also converter) for myself, so some of the initial design decisions were based on what was good enough for me, easier to implement and fit my preferences and files I use. Now that more people use it I'm always open to suggestions to make it useful for as many as possible as long as I feel they improve the program, am able to implement them and fit with the look and feel I'm going for.

That's correct, with the folders I just search for all the images and then bundle them together, mostly to make things easier for myself :) and the files cleaner. I thought that could be a problem only in some rare cases so I went with the more straightforward way, but I'm open to explore doing it differently if altering the folder structure causes problems to some users or has some negative side effects I haven't thought about.

Also by design only image files are included.  I didn't know some people or programs included a ComicInfo.xml file inside comic archives (cbz and cbr ones I'm assuming?) as I've never encountered one or read about it. Is that a widely used thing? (from a first quick search it seems it's a practice initiated by a program called ComicRack). What are its main uses/benefits? And, by the way, does altering the folder structure disturb in some way the validity of the info inside that xml file (if it includes paths to images, for example, that could be the case).

If it's something people would find useful I have no problem trying to add support for ComicInfo.xml in a future version, at the very least to keep it during the conversion. And if I find some kind of standard about its contents (during my quick search I found [this](https://anansi-project.github.io/docs/comicinfo/documentation)) and/or some examples comics that include one, I may even try to add some kind of menu entry to display a window showing the relevant info it contains... As I said, I'm not familiar with it so I'm not really sure what it's commonly used for.

**Comment by 5h4k3** [comment]
No, lol, that's me thanking you for the app, super practical! In fact, I really love that everything is cleared in the archive to only keep images in the final result. I'm a bit of a maniac^^ and I find it's very cool.

Otherwise, changing the structure of the archive has no impact on the inner content of the ComicInfo.xml, there is no PATH in it. It just contains information about the book that are parsed by applications to be displayed.

Since Cbx are basically just archives, there is no mean to embed metadata (writer, penciller, genre, weblink, etc) in them like we can with a Epub for example without create a file with these info next to it or put them in a database of the library manager ([Calibre](https://calibre-ebook.com/) app way). But with this solution, if you send the comic to, let's say, a tablet, all these information will stay on the computer. Sad.

So years ago, two projects attempted to create a standard to handle this issue. 
ComicBookLover on mac, putted these data in the "comment" part of the zip itself but if i recall correctly, it can't handle a lot of information and doesn't work with Cbr (rar). 
And ComicRack created the famous ComicInfo.xml which is placed directly in the archive, with the images and work with every Cbx.

Since then, many desktops and mobile applications are able to read this file to present them elegantly in them UI. And even if ComicRack is not maintained anymore, since no application come as close in term of functionalities, we continue to use it. And like i said, ComicInfo.xml became de facto a kind of standard to hold metadata between devices and new projects like Komga, Kavita or Codex have plan [to (maybe) improve it.](https://github.com/anansi-project/comicinfo) Every new project for managing, self-hosting and cloud delivery or anything related to digital picture books supports it these days since no one has found a better way yet.

So there you go, I think I said it all. And I really think I can speak for all the big comic book consumers and collectors, that, yes, we find this little thing it very useful and we really appreciate it. We spend a lot of time to retrieve these information on the web with the help of scrapper scripts or by hand to have a nice and shinny library.

["Standardization"](https://github.com/anansi-project/comicinfo) (Same link as above). And I drop you a example from one of my book: [ComicInfo.zip](https://github.com/binarynonsense/comic-book-reader/files/11621661/ComicInfo.zip)

Have a nice day !

**Comment by binarynonsense** [comment]
Great! thanks for all the details, now I have enough info to understand why this file may be quite important to some users and start thinking about what I need to do. I don't know when I'll have the time, but this is definitely going in my TODO list and I'll try to add support for ComicInfo.xml in a future version. 

At the very least I'll start by copying the original ComicInfo.xml, if there's one, into the converted file if the format allows it (cbz/r/7), but I may also try to parse it to modify it if needed. From the example you included (thanks for that!) and the proposed standards, there's at least one section that may need updating after a conversion: if the images are resized, or their format changed, each Page section in the Pages section of the xml may need to be updated to reflect the new file size and/or dimensions (seems like the values are in bytes and pixels respectively).

**Comment by 5h4k3** [comment]
You are awesome! Thanks! If like you said, you can just copy it, it would be very nice. 

Parsing the file to update its information and/or to be able to modify it by hand after the conversion would be a great addition without a doubt, of course but I think it's not as essential that to just have it copied back. Not to mention the additional workload for you.

Because, at least in my process, size, dimensions and pages number are calculated by Comicrack itself, this information is not retrieved on the web, unlike all the others, hence the need of copy in the first place.
So, I'm not sure about others software but when the converted file come back in my library, Comicrack is able to see it has change and it will be able to update it by itself.

Anyway, thanks for your time and I wish you success in all your other projects !

**Comment by binarynonsense** [comment]
Unexpectedly, I managed to get some free, quiet time to work on this today, so I've released a [beta version](https://github.com/binarynonsense/comic-book-reader/releases/tag/v2.4.5-beta1) that adds support for ComicBook.xml files to the comic book converter.

I've taken a conservative approach: right now if the input comic contains a ComicBook.xml file and the output format is compatible (cbz or cb7), the file is copied, and ONLY if the images' format or scale is changed I update the Pages section of the xml file to reflect the new values (size, width and height). So if the original xml page data is inaccurate (in number of pages, width...) but there are no image changes during the conversion it will still be inaccurate (I could always update it, but decided to only do it when I was sure there would be changes)... Also, if the xml is invalid or I don't manage to parse it, I just copy it over.

I didn't have more time and don't have any real comic book files that use ComicBook.xml, so my tests have been limited in time and scope (hence the beta release). Please, if you have the time, try checking that everything works as intended using real examples and let me know if that's the case or there are any problems, bugs...

PS: thinking about more ComicBook.xml related improvements for the future, would it be interesting to add a view>comic info (or something like that) menu entry that would open a simple window showing some of the info in the xml file (if available), like the title, authors... things like that? Any other interesting thing other readers use it for or you'd want to have? (no promises on when or if I'll implement any of this, but I'd at least think about it).

**Comment by 5h4k3** [comment]
You rock ! Love your pseudo... You said that you don't know when you will find the time... Few hours later... Beta ! What a non sense !^^

I'm in the process to entirely convert my library in AVIF, 2TB of data. This is going to take ages lol... But it's worth for my backup strategy and that mean more books on my tablet ! So I think I will have the time to hardy test this... And do reports.

For now, I think I have push my computer to its limits, I spin Waifu2x-Extention-Gui with 2 graphics cards for upscaling and next to this, I convert some others with ACBR by list/packs of 50/100 comics. So far, so stable: 100% processor load, same for the nvidia and the amd cards during hours, no freeze, no crash, everything run fine. 

Before using ACBR, I used [CbxConverter](https://tomeko.net/software/CbxConverter/) too and that make me think, it was convenient to select a folder and take recursively every archives in it, add them to the converter's list and recreate the source directory structure afterward (it's gadget imo, but since you ask...). 

And I'm used to work on my books with a ramdisk to avoid too much fragmentation and useless writing with a lot a delete/zip/unzip and I was wondering where ACBR had its Temp folder. Being able to choose a custom one can be nice too. I tend to put every Tmp related to my converters on this thing when I'm able. Especially when downloading, torrenting, converting and upscaling in the same time... Having less load on my disks becomes a goal for stability and piece of mind when going outside or to sleep with all these things running.

And since you ask, maybe you can take a look at [Comicbook Archive Toolbox](https://github.com/degoedel/ComicbookArchiveToolbox), I'm not familiar with licenses stuff but on his repo it say: "under MIT license (do what you want with it)", It's "simple" tool that achieve metadata edition and archives splitting/Merging (can be a nice addition too).

I know, I've more suggested about the converter part than the reader itself but it's my use of ACBR and right now I don't see any other features that I might need, so... voilà...

Anyway, I'll keep you posted on any issues I may encounter. Thanks for all of this.

**Comment by binarynonsense** [comment]
No problem! I haven't had much time to work on things like ACBR lately, only on small updates sometimes, but yesterday was a rare day were things were optimal for a coding session and adding support for ComicBook.xml seemed something I would have time to do and would be useful to some users :)

To create temporary folders/files I use a function in node, os.tmpdir(), that "returns the operating system's default directory for temporary files". In linux seems like it's "/tmp", I'm not in a windows machine right now and don't remember what's the default one there, but you can see the paths to the temp folders ACBR creates during a conversion if you launch it from the command line. I'll think about adding some kind of configuration option to set a custom path (maybe not in the UI for now but available in the config file so advanced users can set it manually), I'll have to change some things but it should be doable without much problem.

I'll also think about your other suggestions, I may try to implement my own metadata viewer/editor someday and maybe split/merge tools (sounds like something that would be more fun and easier to do myself than trying to figure out and adapt other person's code).

**Comment by 5h4k3** [comment]
Ok, I think, I tested it enough and it doesn't seems to have any problem. ComicInfo.xml is always preserved from the the conversion. All fine. 

The only thing I haven't tested much is the automatic update of the ComicInfo.xml after the conversion. I will post back if I find something wrong in the future.

ACBT is currently running on 390 Books with a RamDisk for the temp folder and that works great ! Thanks for this one again! ComicInfo is preserved too... 💯

**Comment by binarynonsense** [comment]
@5h4k3 Hi, I just released a new version of ACBR that brings a big internal and external redesign and I'm now starting to think about expanding the features. First thing I've started experimenting with is a tool to edit/view the ComicInfo.xml inside a file (later on I'll probably add some of the other things we talked about, like the temp folder or the merge/split capability). 

As you are the one that made me aware of their existence and the only person I know that uses them frequently, I was hoping you could help me with some of the decisions I need to make :)
The entries I was thinking of supporting are the ones from the anansi-project schema for v2.0
https://github.com/anansi-project/comicinfo/blob/main/schema/v2.0/ComicInfo.xsd
Would that be the right choice?

Also, I want to divide them in 4 sections, to make them more easy to go through when using the editor/viewer. One for the basic info, the ones most commonly used; another for the creators; one for the pages (including the number of pages and each page's info); and another for the rest. Here is how I'm currently thinking of dividing them, let me know if you think I should switch any to other of the sections (or any other suggestion related to this): 

-- BASIC INFO--
Title (any text)
Series (any text)
Number (any text)
Count (restricted to numbers)
Volume (restricted to numbers)
AlternateSeries (any text)  -> move to others??
AlternateNumber (any text)  -> move to others??
AlternateCount (restricted to numbers)  -> move to others??
Summary (any text)
Notes (any text)  -> move to others??
Year (restricted to numbers)
Month (restricted to numbers)
Day  (restricted to numbers)
Genre (any text)
Format (any text)
ScanInformation (any text)  -> move to others??
LanguageISO  (restricted to two letter iso language ids)

-- CREATORS --
Writer (any text)
Penciller (any text)
Inker (any text)
Colorist (any text)
Letterer (any text)
CoverArtist (any text)
Editor (any text)
Publisher (any text)
Imprint (any text)

-- PAGES --
PageCount (restricted to numbers)
And each page's info

-- OTHER--
Web (any text)
BlackAndWhite  (restricted to yes/no)
Manga  (restricted to special list of choices)
Characters (any text)
Teams (any text)
Locations (any text)
StoryArc (any text)
SeriesGroup (any text)
AgeRating (any text)
CommunityRating (any text)
MainCharacterOrTeam (any text)
Review (any text)

**Comment by 5h4k3** [comment]
Hi !

I've seen you release a fresh version with a new design, but I haven't
tested it yet. And, I just received your mail with the 3.1.0 Beta1, I will
test later tonight.

For the ComicInfo.xml, yes, schema v2.0 seems to be what is in use today.
From what I've seen, some infos that I use are missing in v1.0.

And for the design, I'm not sure I'm the best one to ask, I primarily use
ACBR/ACBT for its conversion fonction, I find it easy to understand, stable
and easy to use. To manage and read them, I tend to use the good old
comicrack with all its plugins.

But, speaking about design without taking the time to take a look at first
seems futile. I will try to test it later today and I will let you know if
I saw something.

But maybe:

ScanInformation (any text ) -> Others (I use it to store resolution like
"1200x1920" but I know some other use it to store scanner nickname info
like "ScanTeamOrion")

But IMO, StoryArc, SeriesGroup,  Alternate Series are for a lot of people
extremely useful and primary information. They need to be not that far from
the title or series name. AlternateCount and AlternateNumber too.
Like how to respect release dates and chronology in the time of a series?
With these alternates numbering for example. Same thing with grouping
series, story arc and co. When reading Marvel or DC universes, these
options become vital.
Alternate Serie or title? A lot of manga have different names. Actually,
minimum 3: Native (進撃の巨人), English (Attack on Titan) or Romaji (Shingeki no
Kyojin).

See you later


CF some screenshots from ComicRack.




Le lun. 14 août 2023 à 18:22, Álvaro García ***@***.***> a
écrit :

**Comment by binarynonsense** [comment]
Thanks for taking the time to write back! Let me know what you think if you try the beta.
Well, you are the only person I know that uses them, so I'll appreciate any tests you can make and any input you can give me :P but I totally understand if you don't have much time to test the beta.
Some of the fields been distributed in different tabs doesn't mean they are less important, just a way to separate them to make all more accessible. But if once you've seen how things are distributed now you feel moving some of those from others to the basic data tab would make them easier to use let me know.
I'm not sure if you mean that you attached some screenshots from ComicRack, but Github must ignore attachment from emails because I can't see any images in your reply (it's probably better to use the web interface for that).

PS: this version also has the customizable temp folder option you wanted.

**Comment by binarynonsense** [comment]
For the field distribution, maybe this will help not wasting too much of your time: I'll list here the current distribution I ended up with, and if after trying the beta you feel some fields should be reordered or moved to another tab, write me back with your modified list:

-- BASIC DATA TAB
Title
Series
Number
Count
Volume
Summary
Year
Month
Day
Genre
Language
-- CREATORS TAB
Writer
Penciller
Inker
Colorist
Letterer
Cover Artist
Editor
Publisher
Imprint
-- PAGES
here goes the pages info and count, this is ok here :)
-- OTHER
Format
Notes
Alternate Series
Alternate Number
Alternate Count
Web
Black and White
Manga
Scan Information
Characters
Main Character or Team
Teams
Locations
Story Arc
Series Group
Age Rating
Community Rating
Review

**Comment by binarynonsense** [comment]
Today I've been trying to find a better distribution and I think I've found a way to show everything in a more compact and useful way:

![image](https://github.com/binarynonsense/comic-book-reader/assets/8535921/96fff9d9-fe1f-4702-b99f-b48b0cf8d6ac)

![image](https://github.com/binarynonsense/comic-book-reader/assets/8535921/24312020-8e3d-4f41-805c-5f07ac437ed1)

![image](https://github.com/binarynonsense/comic-book-reader/assets/8535921/67fc3826-482a-427d-893f-f13d73411161)

![image](https://github.com/binarynonsense/comic-book-reader/assets/8535921/6e119e68-9fa0-48c1-8856-ee67322ac7b4)

**Comment by binarynonsense** [comment]
>  and archives splitting/Merging (can be a nice addition too).
>

Just released [a version](https://github.com/binarynonsense/comic-book-reader/releases/tag/v3.2.1) that lets you split/merge comic files (there's and option int the convert to select the number of files to split the output into, and the create tool now admits comic files as input, so you can just add two or more as input, plus some images if you want, and merge them into one output file).

With this and the metadata editor (that now also lets you search and import data from comic vine, nothing fancy but could be helpful) I think I've added all the things I planned to add some day after our conversation in this issue. Hope you like them :)

**Comment by 5h4k3** [comment]
Hey !  Sorry, just back and got some time to speed in front of the PC. So, after testing It's not like I'm lazy or anything to modify your list. But  informations in the ComicInfo's viewer/Editor are just at a good place. For real, very nice.

I saw the ComicVine's API key slot... You plan to add a scrapper?

Otherwise, I succeed in splitting files but I was not able to merge like 3 CBZ in One... Something I missed?

Keep the good work ! Thanks for It nice little soft.

**Comment by binarynonsense** [comment]
You can use the search section in the comicinfo.xml editor to find data in comic vine (if you've set an api key) and import it to the corresponding fields.

To merge 3 files use the create comic book tool, add them to the list, select the output format and press create.

**Comment by 5h4k3** [comment]
Ha, ok... ! I was not in the right place... Good ! It became my new tool... !

Comicvine search is a great addition too, great work !

And thank you for the choice of the temp folder. I will try if GUI it is as stable as the CLI (I am used to launch 1000+ convert in a shot).

**Comment by binarynonsense** [comment]
Thanks! Glad you like it.
